src/logid/InputDevice.cpp

```diff
@@ -64,6 +61,17 @@ InputDevice::~InputDevice()
     libevdev_free(device);
 }
 
+void InputDevice::registerKey(uint code)
+{
+    libevdev_enable_event_code(device, EV_KEY, code, nullptr);
+    int err = libevdev_uinput_create_from_device(device,
```

Comment:
Nice idea to register the keys on demand (instead of having to parse the config) 😄 

I think we're leaking at least the `ui_device` here. I guess the safe logic would be something like:

1. Check whether we've already enabled the key. If yes, then there's nothing else to do.
2. Otherwise, close the current uinput device.
3. Enable the key on our evdev device.
4. Create a new uinput device.

To be exception-safe, we probably also have to prepare for the case in the `~InputDevice` destructor where the current `ui_device` was already closed, but we haven't created a new one yet.

Reply:
I implemented the changes, thank you.

I think we do not need to check if the input device was closed, since this should be checked by `libevdev`:
https://gitlab.freedesktop.org/libevdev/libevdev/-/blob/master/libevdev/libevdev-uinput.c#L456

Reply:
I'm probably missing something, but don't we still need to set `ui_device` to `nullptr` when it's invalid (because that's what `libevdev-uinput.c` is checking for)? Otherwise, if we throw in https://github.com/PixlOne/logiops/pull/183/commits/8280bc2505295c11f034de9230ef68d80e390785#diff-d5fab82e3435b3bb688775a4b1d9b8778033e00dcd9b192dcaa39b742f7bff3fR92, `libevdev_uinput_destroy` would try to access `uinput_dev->fd` from an already freed pointer when we call it in the destructor.

Reply:
You are right, somehow I thought this is done somewhere, but of course not. It has been too long since I last worked in C/C++ :-)

I think this is also needed in the constructor for `device`, right?
Or am I missing something (again)?

Reply:
If the constructor fails, then [the destructor won't be called](https://isocpp.org/wiki/faq/exceptions#selfcleaning-members), so I think we're fine there.

Not as if this were _that_ important, since if we throw anywhere, the only sensible thing to do is to exit logid, which will clean up the resources, anyways. 🙂

Reply:
Ah I see, thanks for the reference.

Then everything should be cleaned up now and I cannot see any more "problems" with this PR.